tests/helpers/txparser/transaction.py

```diff
@@ -0,0 +1,404 @@
+
+
+class TxParse:
+    """
+    Bitcoin and Bitcoin-derived raw transaction parser.
+
+    Usage:
+
+        - Parse the raw tx into a Python TypedDict object:
+            ``parsed_tx = TxParse.from_raw(raw_btc_tx)``
+    """
+
+    # pylint: disable=too-many-statements
+    @classmethod
+    def from_raw(cls,
+                 raw_tx: Union[bytes, str],
+                 endianness: byteorder = 'little') -> Tx:
+        """
+        Returns a TX object with members initialized from the parsing of the rawTx parameter
+
+        :param raw_tx: The raw transaction to parse. Supported transactions types are:
+            Bitcoin, Bitcoin Segwit, Zcash
+
+        :param endianness: The endianness of values in the raw tx among 'little' or 'big'.
+            Defaults to 'little' (i.e. BTC & derivatives).
+
+        :return: A Tx class (of type TypedDict) with all members initialized.
+        :raise ValueError: If the transaction is malformed or is of an unsupported type.
+        """
+
+        # Internal utilities
+        def _hash(tx: Union[Tx, bytes], show_hashed_items: bool = False) -> str:
+            """Double SHA-256 hash a raw tx or a parsed tx. """
+            def _recursive_hash_obj(obj: Any,
+                                    hasher: Any,
+                                    ignored_fields: Union[List, Tuple],
+                                    path: list,
+                                    show_path: bool = False) -> None:
+                """Recursive hashing of all significant items of a composite object.
+                This inner function is written in a way could be made to an independent one,
+                able to hash the content of any composite dataclass or dict object."""
+                if obj is not None and not isinstance(obj, (bytes, bytearray)):
+                    # Each items in a list of objects must be parsed entirely
+                    if isinstance(obj, list):
+                        for i, item in enumerate(obj):
+                            path.append(str(i + 1))   # Display the item rank in the list
+                            _recursive_hash_obj(item, hasher, ignored_fields, path, show_path)
+                            path.pop()
+                    else:
+                        # Recursively descend into object
+                        attrs = list(obj.__dict__.items())
+                        for key, value in attrs:
+                            # Ignore fields that shan't be hashed => explicitly test for segwit types
+                            # pylint: disable=C0123
+                            if key not in ignored_fields and value is not None and \
+                                    type(value) not in (SegwitExtHeader, SegwitExtFooter):
+                                tmp = path[:]
+                                tmp.append(key)
+                                _recursive_hash_obj(getattr(obj, key), hasher, ignored_fields, tmp, show_path)
+                else:
```

Comment:
in this else, `obj` could be None. Is that your intention? If not and obj should never be None, let's remove the `obj is not None` in the test above.

Reply:
This test fr None is needed. Rationale:
Because the `Tx` class (representing a BTC-compatible transaction parsed to its fields) can have uinitialized fields, I need to ignore those fields when recursively browsing its attributes for fields to hash. For instance, a legacy BTC tx won't have a witness field, so the `SegwitFooterExt`-type `footer` attribute's value will be None for a valid reason. If I don't test for None at the beginning of the function, then it could be recursively called with its parameter `obj` set to None and raise an exception while the parsed tx is valid.

Reply:
ok, got it.
But my issue was more that if indeed obj can be None, then the code below will always fail

```python
In [9]: sha256().update(cast(bytes, None))                                                                                                                                                                  
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-6b2faaacb6e9> in <module>
----> 1 sha256().update(cast(bytes, None))

TypeError: object supporting the buffer API required
```

Reply:
Got it. Changed 
```else: hash.update(cast(bytes, obj))``` 
statement to 
```elif isinstance(obj, (bytes, bytearray)): hasher.update(cast(bytes, obj))```
so that `None` objects are not hashed. But I think this cannot actually occur, even with a `tx` intentionally set to None in the caller test.